NFTBrowser/ImageDownloader.swift

```diff
@@ -0,0 +1,123 @@
+import Foundation
+import UIKit
+import os
+
+class ImageDownloader {
+            
+    let inMemoryCache: InMemoryCache = {
+        return InMemoryCache()
+    }()
+    
+    let session: URLSession = {
+        let config = URLSessionConfiguration.default
+        config.requestCachePolicy = .returnCacheDataElseLoad
+        return URLSession(configuration: config)
+    }()
+    
+    enum ImageError: Error {
+        case loadError
+        case dataError
+        case networkError(Error)
+        case thumbnailError
+    }
+        
+    func fetchImage(url: URL, completion: @escaping ((Result<UIImage, ImageError>)) -> Void) {
+        if let cachedImage = inMemoryCache[key: url] {
+            os_log("Return image from cache %s", log: Log.imageCache, type: .default, url.absoluteString)
+            completion(.success(cachedImage))
+            return
+        }
+        session.dataTask(with: url) { [weak self] data, _, error in
+            if let error = error {
+                DispatchQueue.main.async {
+                    os_log("Network error %s", log: Log.imageCache, type: .error, error.localizedDescription)
+                    completion(.failure(.networkError(error)))
+                }
+                return
+            }
+            guard let data = data else {
+                DispatchQueue.main.async {
+                    os_log("Data error fetching image %s", log: Log.imageCache, type: .error, url.absoluteString)
+                    completion(.failure(.dataError))
+                }
+                return
+            }
+            guard let image = data.thumbnail else {
+                DispatchQueue.main.async {
+                    os_log("Failed to decompress image %s", log: Log.imageCache, type: .error, url.absoluteString)
+                    completion(.failure(.thumbnailError))
+                }
+                return
+            }
```

Comment:
Move it to function `image(with data, error) -> Result<Image>`
let result = image(data, error)

DispatchQueue.main.async {
    completion(result)
}

Reply:
Or, 
```
func image(with urlSessionResult: Result) -> Result<Image> {
// your logic and error handling
}